sequtils: fix double-evaluation bug with `applyIt`

zerbina opened the pull request:
## Summary

Use the `mitems` iterator for implementing the template, fixing the bug
where the argument expression passed to the template was evaluated
multiple times.

The change also makes the template more general: the only requirement
for the input now is that it supports being iterated over via
`mitems`; it no longer needs to be index-able by an integer.

saem commented on tests/lang_callable/template/twrongmapit.nim:
```diff
@@ -30,3 +30,15 @@ import sequtils
 # now works:
 echo "##", i[0], "##"
+
+block evaluation_order:
```
is it worth checking this invariant for the other `xxxit` templates?

zerbina replied:
Yep, it is, but this PR is only meant to fix immediate bug blocking #714. I'd make extending the test case a follow-up PR.

zerbina committed:
sequtils: fix double-evaluation bug with `applyIt`

Summary
=======

Use the `mitems` iterator for implementing the template, fixing the bug
where the argument expression passed to the template was evaluated
multiple times.

The change also makes the template more general: the only requirement
for the input now is that it supports being iterated over via
`mitems`, it no longer needs to be indexable by an integer.

zerbina committed:
Merge branch 'devel' into sequtils-fix-applyIt

zerbina commented:
/merge
